inject/src/main/java/io/micronaut/context/DefaultBeanContext.java

```diff
@@ -2657,43 +2661,63 @@ private void readAllBeanDefinitionClasses() {
                 }
                 if (!hasNonSingletonCandidate) {
                     allCandidatesAreSingleton = true;
                 }
-                beans = beansOfTypeList;
+                beanRegistrations = beansOfTypeList;
             } else {
                 if (LOG.isDebugEnabled()) {
                     LOG.debug("Found no possible candidate beans of type [{}] for qualifier: {} ", beanType.getName(), qualifier);
                 }
                 allCandidatesAreSingleton = true;
-                beans = Collections.emptySet();
+                beanRegistrations = Collections.emptySet();
             }
 
-            if (beans != Collections.EMPTY_SET) {
-                if (Ordered.class.isAssignableFrom(beanType)) {
-                    beans = beans.stream().sorted(OrderUtil.COMPARATOR).collect(StreamUtils.toImmutableCollection());
+            if (beanRegistrations != Collections.EMPTY_SET) {
+                if (Ordered.class.isAssignableFrom(beanType) || hasOrderAnnotation) {
+                    beanRegistrations = beanRegistrations.stream().sorted(BEAN_REGISTRATION_COMPARATOR).collect(StreamUtils.toImmutableCollection());
                 } else {
-                    beans = Collections.unmodifiableCollection(beans);
+                    beanRegistrations = Collections.unmodifiableCollection(beanRegistrations);
                 }
             }
 
+            Collection<T> beans = beanRegistrations.stream().flatMap(b -> Stream.of(b.getBean())).collect(StreamUtils.toImmutableCollection());
             if (allCandidatesAreSingleton) {
                 initializedObjectsByType.put(key, (Collection<Object>) beans);
             }
-            if (LOG.isDebugEnabled() && !beans.isEmpty()) {
+            if (LOG.isDebugEnabled() && !beanRegistrations.isEmpty()) {
                 if (hasQualifier) {
-                    LOG.debug("Found {} beans for type [{} {}]: {} ", beans.size(), qualifier, beanType.getName(), beans);
+                    LOG.debug("Found {} beans for type [{} {}]: {} ", beanRegistrations.size(), qualifier, beanType.getName(), beanRegistrations);
                 } else {
-                    LOG.debug("Found {} beans for type [{}]: {} ", beans.size(), beanType.getName(), beans);
+                    LOG.debug("Found {} beans for type [{}]: {} ", beanRegistrations.size(), beanType.getName(), beanRegistrations);
                 }
             }
 
             return beans;
         }
     }
 
+    private static final Comparator<BeanRegistration> BEAN_REGISTRATION_COMPARATOR = (o1, o2) -> {
+        int order1 = getOrder(o1);
+        int order2 = getOrder(o2);
+        return Integer.compare(order1, order2);
+    };
+
+
+    private static <T> int getOrder(BeanRegistration<T> registration) {
+        Optional<AnnotationValue<Order>> annotation = registration.beanDefinition.findAnnotation(Order.class);
```

Comment:
Please use `beanDefinition.intValue(Order.class)` to obtain the order as an `OptionalInt`